File: run_experiments.js
/**
 * run_experiments.js — Run all 3 HW7/HW8 experiments
 * Uses claude-3-5-haiku for cost efficiency (~$0.01/negotiation)
 *
 * Experiment 1: RFP vs Reverse Auction (6 vendors)
 * Experiment 2: Constraint Propagation Latency at scale (3, 10, 30 vendors)
 * Experiment 3: Failure Recovery (inject failure, measure recovery)
 *
 * Usage: node run_experiments.js [1|2|3|all]
 */

const fs = require("fs");
const path = require("path");
const { runMultiNegotiation, runMultiNegotiationParallel, loadRegistry, MODEL_EXPERIMENT } = require("./negotiate-multi");

const RESULTS_DIR = path.join(__dirname, "results");

// Default form data for experiments
const DEFAULT_FORM = {
  budget: 25000,
  flexBudget: 3000,
  guestCount: 120,
  serviceStyle: "plated",
  cuisinePreference: "Italian-American",
  dietaryNeeds: ["vegetarian", "gluten-free"],
  dateFlexibility: "month",
  altStyles: ["buffet", "family_style"],
  minGuestCount: 100,
  altVenues: [],
};

// Vendor subsets
const SIX_VENDORS = [
  "venue_01", "catering_01", "florist_01",
  "photographer_01", "dj_01", "bakery_01",
];

const TEN_VENDORS = [
  "venue_01", "venue_02",
  "catering_01", "catering_02",
  "florist_01", "florist_02",
  "photographer_01", "photographer_02",
  "dj_01", "bakery_01",
];

const THIRTY_VENDORS = null; // null = all vendors in registry

function ensureDir(dir) {
  if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true });
}

function writeCSV(filePath, headers, rows) {
  const csv = [headers.join(","), ...rows.map(r => r.join(","))].join("\n");
  fs.writeFileSync(filePath, csv);
  console.log(`  📊 Wrote ${filePath}`);
}

// ── Experiment 1: RFP vs Reverse Auction ────────────────────────────────

async function runExperiment1() {
  console.log("\n═══════════════════════════════════════════");
  console.log("  EXPERIMENT 1: RFP vs Reverse Auction");
  console.log("═══════════════════════════════════════════\n");

  const expDir = path.join(RESULTS_DIR, "exp1_rfp_vs_auction");
  ensureDir(expDir);

  // Run RFP mode
  console.log("  🔄 Running RFP mode with 6 vendors...");
  const rfpResult = await runMultiNegotiation(
    { ...DEFAULT_FORM },
    SIX_VENDORS,
    { model: MODEL_EXPERIMENT, mode: "rfp" }
  );
  console.log(`  ✅ RFP complete: ${rfpResult.summary.elapsed_sec}s, $${rfpResult.summary.total_api_cost_usd} API cost`);

  // Run Auction mode
  console.log("  🔄 Running Auction mode with 6 vendors...");
  const auctionResult = await runMultiNegotiation(
    { ...DEFAULT_FORM },
    SIX_VENDORS,
    { model: MODEL_EXPERIMENT, mode: "auction" }
  );
  console.log(`  ✅ Auction complete: ${auctionResult.summary.elapsed_sec}s, $${auctionResult.summary.total_api_cost_usd} API cost`);

  // Generate CSV
  const headers = ["mode", "vendor_id", "vendor_name", "category", "outcome", "total_price", "turns", "api_cost"];
  const rows = [];

  for (const r of rfpResult.results) {
    rows.push(["rfp", r.vendor_id, r.vendor_name, r.category, r.outcome,
      r.final_terms?.total_price || "", r.turns, r.total_cost_usd]);
  }
  for (const r of auctionResult.results) {
    rows.push(["auction", r.vendor_id, r.vendor_name, r.category, r.outcome,
      r.final_terms?.total_price || "", r.turns, r.total_cost_usd]);
  }
  writeCSV(path.join(expDir, "exp1_results.csv"), headers, rows);

  // Summary CSV
  const summHeaders = ["mode", "vendor_count", "accepted", "rejected", "grand_total", "elapsed_sec", "api_cost"];
  const summRows = [
    ["rfp", rfpResult.summary.vendor_count, rfpResult.summary.accepted_count,
      rfpResult.summary.rejected_count, rfpResult.summary.grand_total_wedding_cost,
      rfpResult.summary.elapsed_sec, rfpResult.summary.total_api_cost_usd],
    ["auction", auctionResult.summary.vendor_count, auctionResult.summary.accepted_count,
      auctionResult.summary.rejected_count, auctionResult.summary.grand_total_wedding_cost,
      auctionResult.summary.elapsed_sec, auctionResult.summary.total_api_cost_usd],
  ];
  writeCSV(path.join(expDir, "exp1_summary.csv"), summHeaders, summRows);

  // Save full logs
  fs.writeFileSync(path.join(expDir, "rfp_full.json"), JSON.stringify(rfpResult, null, 2));
  fs.writeFileSync(path.join(expDir, "auction_full.json"), JSON.stringify(auctionResult, null, 2));

  console.log("\n  📋 Experiment 1 Summary:");
  console.log(`     RFP grand total:     $${rfpResult.summary.grand_total_wedding_cost}`);
  console.log(`     Auction grand total:  $${auctionResult.summary.grand_total_wedding_cost}`);
  console.log(`     RFP time:            ${rfpResult.summary.elapsed_sec}s`);
  console.log(`     Auction time:         ${auctionResult.summary.elapsed_sec}s`);
  const savings = rfpResult.summary.grand_total_wedding_cost - auctionResult.summary.grand_total_wedding_cost;
  console.log(`     Auction savings:      $${savings} (${savings > 0 ? "auction cheaper" : "RFP cheaper"})`);

  return { rfp: rfpResult.summary, auction: auctionResult.summary };
}

// ── Experiment 2: Constraint Propagation / Latency at Scale ─────────────

async function runExperiment2() {
  console.log("\n═══════════════════════════════════════════");
  console.log("  EXPERIMENT 2: Scale & Latency");
  console.log("═══════════════════════════════════════════\n");

  const expDir = path.join(RESULTS_DIR, "exp2_scale_latency");
  ensureDir(expDir);

  const scales = [
    { label: "3 vendors", ids: SIX_VENDORS.slice(0, 3) },
    { label: "6 vendors", ids: SIX_VENDORS },
    { label: "10 vendors", ids: TEN_VENDORS },
  ];

  // For HW8, add 30 vendors
  const registry = loadRegistry();
  const allIds = [];
  for (const cat of Object.values(registry.categories)) {
    for (const v of cat.vendors) allIds.push(v.id);
  }
  scales.push({ label: "30 vendors", ids: allIds });

  const scaleResults = [];

  for (const scale of scales) {
    console.log(`  🔄 Running sequential with ${scale.label}...`);
    const seqResult = await runMultiNegotiation(
      { ...DEFAULT_FORM },
      scale.ids,
      { model: MODEL_EXPERIMENT, mode: "rfp" }
    );
    console.log(`  ✅ Sequential ${scale.label}: ${seqResult.summary.elapsed_sec}s`);

    console.log(`  🔄 Running parallel with ${scale.label}...`);
    const parResult = await runMultiNegotiationParallel(
      { ...DEFAULT_FORM },
      scale.ids,
      { model: MODEL_EXPERIMENT, concurrency: 5 }
    );
    console.log(`  ✅ Parallel ${scale.label}: ${parResult.summary.elapsed_sec}s`);

    scaleResults.push({
      vendor_count: scale.ids.length,
      label: scale.label,
      sequential_sec: seqResult.summary.elapsed_sec,
      parallel_sec: parResult.summary.elapsed_sec,
      speedup: Math.round((seqResult.summary.elapsed_sec / parResult.summary.elapsed_sec) * 100) / 100,
      seq_cost: seqResult.summary.total_api_cost_usd,
      par_cost: parResult.summary.total_api_cost_usd,
      seq_tokens: seqResult.summary.total_tokens,
      par_tokens: parResult.summary.total_tokens || 0,
    });

    // Save individual logs
    fs.writeFileSync(
      path.join(expDir, `seq_${scale.ids.length}v.json`),
      JSON.stringify(seqResult, null, 2)
    );
    fs.writeFileSync(
      path.join(expDir, `par_${scale.ids.length}v.json`),
      JSON.stringify(parResult, null, 2)
    );
  }

  // CSV
  const headers = ["vendor_count", "sequential_sec", "parallel_sec", "speedup", "seq_cost", "par_cost", "seq_tokens", "par_tokens"];
  const rows = scaleResults.map(r => [
    r.vendor_count, r.sequential_sec, r.parallel_sec, r.speedup,
    r.seq_cost, r.par_cost, r.seq_tokens, r.par_tokens,
  ]);
  writeCSV(path.join(expDir, "exp2_results.csv"), headers, rows);

  console.log("\n  📋 Experiment 2 Summary:");
  for (const r of scaleResults) {
    console.log(`     ${r.label}: seq=${r.sequential_sec}s, par=${r.parallel_sec}s, speedup=${r.speedup}x`);
  }

  return scaleResults;
}

// ── Experiment 3: Failure Recovery ──────────────────────────────────────

async function runExperiment3() {
  console.log("\n═══════════════════════════════════════════");
  console.log("  EXPERIMENT 3: Failure Recovery");
  console.log("═══════════════════════════════════════════\n");

  const expDir = path.join(RESULTS_DIR, "exp3_failure_recovery");
  ensureDir(expDir);

  // Baseline: no failures
  console.log("  🔄 Running baseline (no failures) with 6 vendors...");
  const baseline = await runMultiNegotiation(
    { ...DEFAULT_FORM },
    SIX_VENDORS,
    { model: MODEL_EXPERIMENT, mode: "rfp" }
  );
  console.log(`  ✅ Baseline: ${baseline.summary.elapsed_sec}s, $${baseline.summary.grand_total_wedding_cost} total`);

  // With failure: venue fails, triggers backup
  console.log("  🔄 Running with venue failure (venue_01 → backup)...");
  const venueFailure = await runMultiNegotiation(
    { ...DEFAULT_FORM },
    SIX_VENDORS,
    { model: MODEL_EXPERIMENT, mode: "rfp", failureVendorId: "venue_01" }
  );
  console.log(`  ✅ Venue failure: ${venueFailure.summary.elapsed_sec}s, $${venueFailure.summary.grand_total_wedding_cost} total`);

  // With failure: caterer fails
  console.log("  🔄 Running with caterer failure (catering_01 → backup)...");
  const cateringFailure = await runMultiNegotiation(
    { ...DEFAULT_FORM },
    SIX_VENDORS,
    { model: MODEL_EXPERIMENT, mode: "rfp", failureVendorId: "catering_01" }
  );
  console.log(`  ✅ Caterer failure: ${cateringFailure.summary.elapsed_sec}s, $${cateringFailure.summary.grand_total_wedding_cost} total`);

  // CSV
  const headers = ["scenario", "vendor_count", "accepted", "failed", "recovered", "grand_total", "elapsed_sec", "api_cost"];
  const rows = [
    ["baseline", baseline.summary.vendor_count, baseline.summary.accepted_count,
      0, 0, baseline.summary.grand_total_wedding_cost,
      baseline.summary.elapsed_sec, baseline.summary.total_api_cost_usd],
    ["venue_failure", venueFailure.summary.vendor_count, venueFailure.summary.accepted_count,
      venueFailure.summary.failed_count,
      venueFailure.results.filter(r => r.is_backup).length,
      venueFailure.summary.grand_total_wedding_cost,
      venueFailure.summary.elapsed_sec, venueFailure.summary.total_api_cost_usd],
    ["catering_failure", cateringFailure.summary.vendor_count, cateringFailure.summary.accepted_count,
      cateringFailure.summary.failed_count,
      cateringFailure.results.filter(r => r.is_backup).length,
      cateringFailure.summary.grand_total_wedding_cost,
      cateringFailure.summary.elapsed_sec, cateringFailure.summary.total_api_cost_usd],
  ];
  writeCSV(path.join(expDir, "exp3_results.csv"), headers, rows);

  // Save full logs
  fs.writeFileSync(path.join(expDir, "baseline.json"), JSON.stringify(baseline, null, 2));
  fs.writeFileSync(path.join(expDir, "venue_failure.json"), JSON.stringify(venueFailure, null, 2));
  fs.writeFileSync(path.join(expDir, "catering_failure.json"), JSON.stringify(cateringFailure, null, 2));

  console.log("\n  📋 Experiment 3 Summary:");
  console.log(`     Baseline total:      $${baseline.summary.grand_total_wedding_cost}`);
  console.log(`     Venue failure total:  $${venueFailure.summary.grand_total_wedding_cost}`);
  console.log(`     Caterer failure total: $${cateringFailure.summary.grand_total_wedding_cost}`);
  const venueOverhead = venueFailure.summary.elapsed_sec - baseline.summary.elapsed_sec;
  console.log(`     Recovery overhead:    ${venueOverhead}s (venue), ${cateringFailure.summary.elapsed_sec - baseline.summary.elapsed_sec}s (caterer)`);

  return { baseline: baseline.summary, venueFailure: venueFailure.summary, cateringFailure: cateringFailure.summary };
}

// ── Main ────────────────────────────────────────────────────────────────

async function main() {
  const arg = process.argv[2] || "all";
  ensureDir(RESULTS_DIR);

  console.log("╔═══════════════════════════════════════════╗");
  console.log("║  ALTARA — A2A Negotiation Experiments     ║");
  console.log("║  Model: claude-3-5-haiku (cost-efficient) ║");
  console.log("╚═══════════════════════════════════════════╝");

  const allResults = {};

  if (arg === "1" || arg === "all") {
    allResults.exp1 = await runExperiment1();
  }
  if (arg === "2" || arg === "all") {
    allResults.exp2 = await runExperiment2();
  }
  if (arg === "3" || arg === "all") {
    allResults.exp3 = await runExperiment3();
  }

  // Save combined results
  const combinedPath = path.join(RESULTS_DIR, "all_experiments_summary.json");
  fs.writeFileSync(combinedPath, JSON.stringify(allResults, null, 2));
  console.log(`\n✅ All results saved to ${RESULTS_DIR}/`);
  console.log(`   Combined summary: ${combinedPath}`);
}

main().catch(err => {
  console.error("❌ Experiment failed:", err.message);
  process.exit(1);
});

File: run_experiments.py
#!/usr/bin/env python3
"""
run_experiments.py — Python wrapper for Altara A2A negotiation experiments
Calls the Node.js experiment runner and generates charts from CSVs.

Usage:
  python3 run_experiments.py          # run all experiments + charts
  python3 run_experiments.py 1        # run experiment 1 only
  python3 run_experiments.py charts   # generate charts from existing CSVs
"""

import subprocess
import sys
import os
import csv

RESULTS_DIR = os.path.join(os.path.dirname(__file__), "results")


def run_node_experiments(exp_num="all"):
    """Run the Node.js experiment runner."""
    print(f"\n🚀 Running experiment(s): {exp_num}\n")
    result = subprocess.run(
        ["node", "run_experiments.js", str(exp_num)],
        cwd=os.path.dirname(__file__) or ".",
        capture_output=False,
    )
    if result.returncode != 0:
        print(f"❌ Node experiments failed with code {result.returncode}")
        sys.exit(1)
    print("✅ Node experiments completed successfully")


def generate_charts():
    """Generate ASCII charts from CSV results (no matplotlib needed)."""
    print("\n📊 Generating result summaries...\n")

    # Experiment 1 chart
    exp1_csv = os.path.join(RESULTS_DIR, "exp1_rfp_vs_auction", "exp1_summary.csv")
    if os.path.exists(exp1_csv):
        print("━━━ Experiment 1: RFP vs Reverse Auction ━━━")
        with open(exp1_csv) as f:
            reader = csv.DictReader(f)
            rows = list(reader)
        for row in rows:
            mode = row["mode"].upper()
            total = float(row["grand_total"])
            bar = "█" * int(total / 500)
            print(f"  {mode:8s} ${total:>10,.0f} {bar}")
        print()

    # Experiment 2 chart
    exp2_csv = os.path.join(RESULTS_DIR, "exp2_scale_latency", "exp2_results.csv")
    if os.path.exists(exp2_csv):
        print("━━━ Experiment 2: Scale & Latency ━━━")
        with open(exp2_csv) as f:
            reader = csv.DictReader(f)
            rows = list(reader)
        print(f"  {'Vendors':>8s} {'Sequential':>12s} {'Parallel':>12s} {'Speedup':>8s}")
        print(f"  {'─' * 8} {'─' * 12} {'─' * 12} {'─' * 8}")
        for row in rows:
            n = row["vendor_count"]
            seq = float(row["sequential_sec"])
            par = float(row["parallel_sec"])
            speedup = float(row["speedup"])
            print(f"  {n:>8s} {seq:>10.1f}s {par:>10.1f}s {speedup:>6.1f}x")
        print()

    # Experiment 3 chart
    exp3_csv = os.path.join(RESULTS_DIR, "exp3_failure_recovery", "exp3_results.csv")
    if os.path.exists(exp3_csv):
        print("━━━ Experiment 3: Failure Recovery ━━━")
        with open(exp3_csv) as f:
            reader = csv.DictReader(f)
            rows = list(reader)
        for row in rows:
            scenario = row["scenario"]
            total = float(row["grand_total"])
            elapsed = float(row["elapsed_sec"])
            failed = row.get("failed", 0)
            recovered = row.get("recovered", 0)
            print(f"  {scenario:20s} total=${total:>10,.0f}  time={elapsed:>6.1f}s  failed={failed} recovered={recovered}")
        print()


def try_matplotlib_charts():
    """Try to generate PNG charts using matplotlib (optional)."""
    try:
        import matplotlib
        matplotlib.use("Agg")
        import matplotlib.pyplot as plt
    except ImportError:
        print("  ℹ️  matplotlib not installed — skipping PNG chart generation")
        print("     Install with: pip3 install matplotlib")
        return

    # Experiment 2: latency bar chart
    exp2_csv = os.path.join(RESULTS_DIR, "exp2_scale_latency", "exp2_results.csv")
    if os.path.exists(exp2_csv):
        with open(exp2_csv) as f:
            reader = csv.DictReader(f)
            rows = list(reader)

        vendors = [r["vendor_count"] for r in rows]
        seq = [float(r["sequential_sec"]) for r in rows]
        par = [float(r["parallel_sec"]) for r in rows]

        x = range(len(vendors))
        fig, ax = plt.subplots(figsize=(8, 5))
        ax.bar([i - 0.2 for i in x], seq, 0.4, label="Sequential", color="#4A90D9")
        ax.bar([i + 0.2 for i in x], par, 0.4, label="Parallel", color="#50C878")
        ax.set_xlabel("Number of Vendors")
        ax.set_ylabel("Time (seconds)")
        ax.set_title("Experiment 2: Negotiation Latency at Scale")
        ax.set_xticks(list(x))
        ax.set_xticklabels(vendors)
        ax.legend()
        chart_path = os.path.join(RESULTS_DIR, "exp2_scale_latency", "exp2_chart.png")
        fig.savefig(chart_path, dpi=150, bbox_inches="tight")
        print(f"  📈 Saved {chart_path}")

    # Experiment 1: comparison bar chart
    exp1_csv = os.path.join(RESULTS_DIR, "exp1_rfp_vs_auction", "exp1_summary.csv")
    if os.path.exists(exp1_csv):
        with open(exp1_csv) as f:
            reader = csv.DictReader(f)
            rows = list(reader)

        modes = [r["mode"].upper() for r in rows]
        totals = [float(r["grand_total"]) for r in rows]
        times = [float(r["elapsed_sec"]) for r in rows]

        fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(10, 4))
        ax1.bar(modes, totals, color=["#4A90D9", "#E8833A"])
        ax1.set_ylabel("Grand Total ($)")
        ax1.set_title("Wedding Cost by Mode")

        ax2.bar(modes, times, color=["#4A90D9", "#E8833A"])
        ax2.set_ylabel("Time (seconds)")
        ax2.set_title("Negotiation Time by Mode")

        fig.suptitle("Experiment 1: RFP vs Reverse Auction")
        chart_path = os.path.join(RESULTS_DIR, "exp1_rfp_vs_auction", "exp1_chart.png")
        fig.savefig(chart_path, dpi=150, bbox_inches="tight")
        print(f"  📈 Saved {chart_path}")


if __name__ == "__main__":
    arg = sys.argv[1] if len(sys.argv) > 1 else "all"

    if arg == "charts":
        generate_charts()
        try_matplotlib_charts()
    else:
        run_node_experiments(arg)
        generate_charts()
        try_matplotlib_charts()
